## Задача 1
###  Условие задачи

В цехах N1 и N2 предприятия производится продукт Y, который в дальнейшем используется
в качестве исходного материала для производства изделий в цехе N3. Суммарная производительность цехов N1 и N2 зависит от вложения дополнительных средств X. При работе цехов N1 и N2 в течение одного месяца эта зависимость может быть приближенно представлена в виде функций:  
* $N1: y = 5+(x+40)^{2/3};$
* $N2: y = 7+(x+30)^{1/2};$  
Функции остатка средств в течение месяца: 
* $N1: 0.7x;$
* $N2: 0.94x;$  
Средства, выделяемые на оба цеха в течение квартала (3 месяца), составляют 174 единиц; перераспределение производится помесячно.  
 Требуется распределить средства на планируемый квартал с целью получения максимального количества продукта Y.
 
 
### Решение
Средства, выделяемые на оба цеха в течение квартала (3 месяца), составляют 174 единиц.    
Перераспределение средств помесячное, поэтому задачу можно разбить на 3 этапа.  
Обозначим главное понятия динамического программирования:  
Выигрыш - максимум произведенных продуктов на обеих цехах N1 и N2.  
Управление - принятие решения по распределению средств.  
Состояние - остаток средств от обеих цехов.  
$ a_{k} = x_{1k}+x_{2k} - остаток средств на производном шаге, z_{k} - суммарное производство N1+N2$  
Для рассчётов использую python 3.7, код:  


In [23]:
#-*-coding: utf-8-*-
"""import mathplot.pyplot as plt"""
"""import operator"""
iteration = 0;
start_funds = 174;
# Призводительность цеха 1
def workshop1_perf(x):
    return (5+(x+40)**(2/3))
# Призводительность цеха 2
def workshop2_perf(x):
    return (7+(x+30)**(1/2))
# Общая производительность за месяц
def full_perf(funds, x):
    return (workshop1_perf(x)+workshop2_perf(funds-x))
# Остаток производства за месяц
def rest(funds,x):
    return (0.7*x+0.94*(funds-x))
# Главная функция 
def magic (funds):
    global iteration;
    global best_x;
    iteration += 1;
    x_list = [i for i in range (int(funds))];
    if (iteration<3):
        next_results_list=[full_perf(funds,i)+magic(rest(funds,i)) for i in x_list];
    else:
        next_results_list = [full_perf(funds,i) for i in x_list];
    best_result=max(next_results_list);
    iteration -= 1;
    best_x =next_results_list.index(best_result);
    return best_result;
for i in range(3):
    result =magic(start_funds);
    if(iteration==0):
           fullresult=int(result);
    iteration+=1;
    print("Period: ");
    print(iteration);
    print("Investing in first: ");
    print(best_x);
    print("Investing in second: ");
    print(int(start_funds - best_x));
                
    start_funds=rest(start_funds,best_x);
print("Total win: ");
print(fullresult);

Period: 
1
Investing in first: 
143
Investing in second: 
31
Period: 
2
Investing in first: 
127
Investing in second: 
2
Period: 
3
Investing in first: 
90
Investing in second: 
1
Total win: 
143


Оптимальный план будет иметь вид:  

| Месяц                               | 1   | 2   | 3  |
|-------------------------------------|-----|-----|----|
| Средства, выделяемые на первом цехе | 143 | 127 | 90 |
| Средства, выделяемые на втором цехе | 31  | 2   | 1  | 

Общее количество продукта будет 143 единицы.
